### PR TITLE
Decision Model Logic with Xunit tests

### PR DESCRIPTION
**Pull request by Kakarot**
I took @tparnell8  feedback and deleted the previous unit tests and instead committed the decision logic with unit tests that tested the concrete class.

The concrete class **AllergyIntolerance** is part of the model. It expects the EHR prototype to reach out to us with the _Patient's ID_ , and list of medications that the EHR provider is asking us to determine if there is an allergy conflict for the patient.

So the **AllergyIntolerance** class is our decision logic that will take in the PatientID and List of Medications, fetch the Patient's list of known allergies, compare their allergies against their medications, and has the capability of returning that their is a problem as well  as list of Medications that will trigger an allergic reaction.

Because the webservice does not use a database, I currently only store one medication/allergy using a dictionary upon instantiation of the class. It is a solid proof of concept and it passes the unit tests which also test for negative cases such as bad medication names, invalid patients, etc. Going forward, the dictionary is easily extensible (currently a one-liner) and should not require modifying any other portion of the code as we support more allergy codes and known medication conflicts.

I tried to minimize duplicate code, there is only one section that somewhat bothers me, but it gets the job done. If the only complaint is style or standards, I'm completely okay with refactoring, but I'm more concerned that the parameters I'm using are exactly what expect to use when making decisions.

### CODE REVIEW

File: src/EhrgoHealth.WebService.UnitTests/TestAllergyIntolerance.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        //Pass in valid medication that a patient has an allergy to and expect a returned list of medication the patient
+        //is allergic to.
+        [Fact]
+        public void TestGetListOfMedicationAllergies01()
+        {
+            Models.AllergyIntolerance allergyIntolerance = new Models.AllergyIntolerance("http://fhirtest.uhn.ca/baseDstu2/");
+            uint patientID = 6140;
+            List<string> medications = new List<string>() { "hydrocodone" };
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService.UnitTests/TestAllergyIntolerance.cs]: its nit picky but writing `List<string> medications = new List<string>...` is quite verbose when you could just write `var medications = new List<string>()`

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    public class AllergyIntolerance
+    {
+        //Dictonary looks up possible allergies for a given medication. Example allergyLookup<Medication, List<Codes>>
+        private Dictionary<string, List<string>> allergyLookup = new Dictionary<string, List<string>>();
+        private FhirClient fhirClient = null;
+
+
+        public AllergyIntolerance(string fhirServer)
+        {
+            if (fhirServer == null || fhirServer == String.Empty)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: `if(String.IsNullOrEmpty(fhirServer))`

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                throw new Exception("Invalid URL passed to AllergyIntolerance Constructor");
+
+            fhirClient = new FhirClient(fhirServer);
+
+            //We do not use a database for the webservice. We will hardcode a handful of allergies and their intolerances
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: this is perfectly valid at this stage of a prototype...mind moving the data to a constants file/class

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        /// <param name="medications"> List of patients medications</param>
+        /// <param name="lookupPatientsKnownAllergies"> Dictonary of patient's known allergies</param>
+        public Dictionary<string, Boolean> GetPatientsKnownAllergies(uint patientID)
+        {
+            var allergyResource = fhirClient.Read<Hl7.Fhir.Model.AllergyIntolerance>("AllergyIntolerance/" + patientID);
+            //Coding coding = allergyResource.Substance.Coding.First<Coding>();
+            if (allergyResource.Substance.Coding.Count == 0)
+            {
+                //Patient has no known allergies, exit early to avoid unnecessary logic execution
+                return null;
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: null pointers typically cause issues in applications as they throw exceptions when accessed as objects...consider returning an empty dictionary instead and have the rest of the code account for an empty dictionary. This will let callers of this code not have to worry about getting a null ref.

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            }
+            List<Coding> patientsKnownAllergies = allergyResource.Substance.Coding.ToList();
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: why the ToList?

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+            //Create a dictionary for O(1) lookup time later.
+            Dictionary<string, Boolean> lookupPatientsKnownAllergies = new Dictionary<string, Boolean>();
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: consider the following code instead

```csharp

return allergyResource.Substance.Coding.ToDictionary(a=>a.code, true)


```

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        /// <param name="medications"> List of patients medications</param>
+        /// <param name="lookupPatientsKnownAllergies"> Dictionary of allergies the patient has, populated from FHIR</param>
+        private List<string> DetermineListOfAllergicMedications(List<string> medications, Dictionary<string, bool> lookupPatientsKnownAllergies)
+        {
+            List<string> listOfAllergicMedications = new List<string>();
+            List<string> currentAllergyCodeList = new List<string>();
+            foreach (var m in medications)
+            {
+                allergyLookup.TryGetValue(m.ToLower(), out currentAllergyCodeList);
+                if (currentAllergyCodeList != null && currentAllergyCodeList.Count > 0)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]:
```csharp

if (currentAllergyCodeList == null || currentAllergyCodeList.Count > 0){
 continue;
}
//add range is faster as it will resize the array only once to accommodate the extra items.
listOfAllergicMedications.AddRange(
currentAllergyCodeList
.Where(c=>lookupPatientsKnownAllergies.ContainsKey(c)))

```

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+        //Populates the list of medications the patient is allergic to
+        /// <summary>
+        /// Private helper method.
+        /// </summary>       
+        /// <param name="medications"> List of patients medications</param>
+        /// <param name="lookupPatientsKnownAllergies"> Dictionary of allergies the patient has, populated from FHIR</param>
+        private List<string> DetermineListOfAllergicMedications(List<string> medications, Dictionary<string, bool> lookupPatientsKnownAllergies)
+        {
+            List<string> listOfAllergicMedications = new List<string>();
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: consider just using var here

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+
+
+        /// Private helper method for comparing the patient's medications against their known allergies
+        /// <summary>       
+        /// I'm the least happy with this method, because I know I can refactor this to do more to also give me a list or dictionary of my results.
+        /// </summary>       
+        /// <param name="medications"> List of patients medications</param>
+        /// <param name="lookupPatientsKnownAllergies"> Dictonary of patient's known allergies</param>
+        private Boolean IsAllergic(List<string> medications, Dictionary<string, Boolean> lookupPatientsKnownAllergies)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: the medications parameter can be an IEnumerable<String>

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                //There are no records on the FHIR server of this patient having any allergies.
+                return false;
+            }
+            return IsAllergic(medications, lookupPatientsKnownAllergies);
+
+        }//end IsAllergicToMedications method
+
+
+
+        /// Private helper method for comparing the patient's medications against their known allergies
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: great use of comments

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs (hunk 9 above)
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: lookupPatientsKnownAllergies should be of type `IDictionary<string, bool>`

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        /// </summary>       
+        /// <param name="medications"> List of patients medications</param>
+        /// <param name="lookupPatientsKnownAllergies"> Dictonary of patient's known allergies</param>
+        private Boolean IsAllergic(List<string> medications, Dictionary<string, Boolean> lookupPatientsKnownAllergies)
+        {
+            List<string> currentAllergyCodeList = new List<string>();
+            foreach (var m in medications)
+            {
+                allergyLookup.TryGetValue(m.ToLower(), out currentAllergyCodeList);
+                if (currentAllergyCodeList != null && currentAllergyCodeList.Count > 0)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]:
```csharp

return 
  currentAllergyCodeList != null 
  && currentAllergyCodeList.Count > 0
  && currentAllergyCodeList.Any(c=>lookupPatientsKnownAllergies.ContainsKey(c))
```

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 12]
```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Rest;
+
+namespace EhrgoHealth.WebService.Models
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: I'd personally not consider this a model, models in web land typically do not really have code, they are more representations of datastructures either in the database, or data passed into/from api calls

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+
+        //The controller is expected to receive two pieces of data from an EHR.
+        //The List of medications they currently know their patient is taking
+        //The patient's ID from FHIR
+        /// <summary>
+        /// The controller is expected to receive two data structures from an EHR.
+        /// </summary>       
+        /// <param name="patientID"> The patient's ID from FHIR</param>
+        /// <param name="medications"> The List of medications they currently know their patient is taking</param>
+        public List<string> GetListOfMedicationAllergies(uint patientID, List<string> medications)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: I'm sure you can make your int's ints and not unsigned int's although I do not predict we will need negative numbers, I also do not predict us needing int's over 2 million

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        {
+            //ToDo: Discuss with team about returning a dictionary or wrapper class, so that we can inform the EHR
+            //      on both the medication conflict and the allergy code.
+
+
+            //First let us fetch the known allergies of the patient.
+            List<string> returnedAllergies = new List<string>();
+            Dictionary<string, Boolean> lookupPatientsKnownAllergies = GetPatientsKnownAllergies(patientID);
+            if (lookupPatientsKnownAllergies == null)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]:
```csharp

return lookupPatientsKnownAllergies == null ? 
    returnedAllergies
    : DetermineListOfAllergicMedications(medications, lookupPatientsKnownAllergies);
```

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+            {
+                //There are no records on the FHIR server of this patient having any allergies.
+                return returnedAllergies;
+            }         
+
+            ////Now we retrieve the list of known allergies their medications can trigger
+            List<string> listOfAllergicMedications = DetermineListOfAllergicMedications(medications, lookupPatientsKnownAllergies);
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: you could have just written `return DetermineListOfAllergicMedications(medications, lookupPatientsKnownAllergies);

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+        //The patient's ID from FHIR
+        /// <summary>
+        /// The controller is expected to receive two data structures from an EHR.
+        /// </summary>       
+        /// <param name="patientID"> The patient's ID from FHIR</param>
+        /// <param name="medications"> The List of medications they currently know their patient is taking</param>
+        public List<string> GetListOfMedicationAllergies(uint patientID, List<string> medications)
+        {
+            //ToDo: Discuss with team about returning a dictionary or wrapper class, so that we can inform the EHR
+            //      on both the medication conflict and the allergy code.
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: sounds legit

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+    {
+        //Dictonary looks up possible allergies for a given medication. Example allergyLookup<Medication, List<Codes>>
+        private Dictionary<string, List<string>> allergyLookup = new Dictionary<string, List<string>>();
+        private FhirClient fhirClient = null;
+
+
+        public AllergyIntolerance(string fhirServer)
+        {
+            if (fhirServer == null || fhirServer == String.Empty)
+                throw new Exception("Invalid URL passed to AllergyIntolerance Constructor");
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]:
```csharp

throw new ArgumentNullException(nameof(fhirServer));

```

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+
+        /// If you only care if a patient is allergic to one of their medications, call this method.
+        /// <summary>
+        /// This method can take use the data given to it from the EHR, as is, and determine if there is a problem.
+        /// However, the downside is this method only returns a Boolean. If you want a list of the medications
+        /// the patient is allergic to, then call GetListOfMedicationAllergies() instead.
+        /// </summary>       
+        /// <param name="patientID"> The patient's ID from FHIR</param>
+        /// <param name="medications"> The List of medications they currently know their patient is taking</param>
+        public Boolean IsAllergicToMedications(uint patientID, List<String> medications)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]:
```csharp

public bool IsAllergicToMedications(...)

```

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+        {
+            //First let us fetch the known allergies of the patient.
+            List<string> returnedAllergies = new List<string>();
+            Dictionary<string, Boolean> lookupPatientsKnownAllergies = GetPatientsKnownAllergies(patientID);
+            if (lookupPatientsKnownAllergies == null)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]:
```csharp

return lookupPatientsKnownAllergies == null ? false : IsAllergic(medications, lookupPatientsKnownAllergies);

```

**Comment by tparnell8** [comment]
There is a lot of little nit picky code issues, but nothing crazy. Try to take IEnumerable/IDictionary as parameters in methods since all arrays inherit from them, unless there is something specific in the List interface you need like Add or AddRange

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+
+        /// Returns a list of patient's allergies from the FHIR server
+        /// <summary>       
+        /// If you want the list of allergy codes of a patient, then just pass in the patient's ID.
+        /// </summary>       
+        /// <param name="medications"> List of patients medications</param>
+        /// <param name="lookupPatientsKnownAllergies"> Dictonary of patient's known allergies</param>
+        public Dictionary<string, Boolean> GetPatientsKnownAllergies(uint patientID)
+        {
+            var allergyResource = fhirClient.Read<Hl7.Fhir.Model.AllergyIntolerance>("AllergyIntolerance/" + patientID);
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: you have a using Hl7.Fhir.Model statement at the top of this file you shouldn't need this here...also refactoring essentials plugin for VS would have grayed this text out.

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs (hunk 6 above)
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: so I'm wondering if you could reduce this method to look like this

```csharp

return allergyResource.Substance.Coding.Count == 0  ?
 null
: allergyResource.Substance.Coding.ToDictionary(a=>a.Code, a=>true)

```

File: src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+            return false;
+        }//end IsAllergic method
+
+        /// Returns a list of patient's allergies from the FHIR server
+        /// <summary>       
+        /// If you want the list of allergy codes of a patient, then just pass in the patient's ID.
+        /// </summary>       
+        /// <param name="medications"> List of patients medications</param>
+        /// <param name="lookupPatientsKnownAllergies"> Dictonary of patient's known allergies</param>
+        public Dictionary<string, Boolean> GetPatientsKnownAllergies(uint patientID)
```
Comment (tparnell8) [inline src/EhrgoHealth.WebService/Models/AllergyIntolerance.cs]: writing out Boolean seems very verbose when you could just write bool

**Comment by Kakarot** [comment]
Hey @tparnell8 , I was looking at the Saturday night's pull request to see if it had been merged, and I see appveyor says that it did not pass 1 out of 2 checks.
commits_1


I went to AppVeyor's build site and the console output says the build passed.
buildsuccess


When I look at the unit tests on appveyor, they also pass. 
screenshots



I downloaded the code onto another computer to check, and everything builds and all the unit test pass there also. 
buildsuccess2


Do you know what the problem is?
